File: WebClient SDK/src/default/tiles/krSettingsTileExtension.ts
import {
  TileExtension,
  ITileLocalExtensionContext,
  Tile,
  TileGroups,
  TileEvaluationEventArgs
} from 'tessa/ui/tiles';
import { userSession } from 'common/utility';
import {
  UIContext,
  UIButton,
  showError,
  showMessage,
  showConfirm,
  showLoadingOverlay,
  showNotEmpty
} from 'tessa/ui';
import { ICardModel } from 'tessa/ui/cards';
import { showFormDialog } from 'tessa/ui/uiHost';
import { DotNetType, createTypedField } from 'tessa/platform';
import { CardSection } from 'tessa/cards';
import { localize } from 'tessa/localization/localize';
import { CardRequest, CardService, CardResponse } from 'tessa/cards/service';

export class KrSettingsTileExtension extends TileExtension {
  public initializingLocal(context: ITileLocalExtensionContext): void {
    const panel = context.workspace.leftPanel;
    panel.tiles.push(
      new Tile({
        name: 'GenerateTestCards',
        caption: '$KrTiles_GenerateTestCards',
        icon: 'ta icon-thin-001',
        contextSource: panel.contextSource,
        command: KrSettingsTileExtension.generateAction,
        group: TileGroups.Cards,
        order: 10,
        evaluating: KrSettingsTileExtension.enableOnSettingsCardAndAdministrator
      })
    );
  }

  private static enableOnSettingsCardAndAdministrator(e: TileEvaluationEventArgs) {
    const editor = e.current.context.cardEditor;

    e.setIsEnabledWithCollapsing(
      e.currentTile,
      !!editor &&
        !!editor.cardModel &&
        editor.cardModel.cardType.id === '35a03878-57b6-4263-ae36-92eb59032132' && // KrSettingsTypeID
        userSession.isAdmin
    );
  }

  private static async generateAction() {
    const editor = UIContext.current.cardEditor;
    let model: ICardModel;
    if (editor && (model = editor.cardModel!)) {
      const section = model.card.sections.get('KrCardGeneratorVirtual')!;
      const form = model.cardType.forms.find(x => x.name === 'Generator')!;

      await showFormDialog(
        form,
        model,
        () => {
          section.fields.set('UserCount', 0, DotNetType.Int32);
          section.fields.set('PartnerCount', 0, DotNetType.Int32);
        },
        [
          new UIButton('$KrTiles_CreateCardsButton', btn =>
            KrSettingsTileExtension.generateButtonAction(section, () => btn.close())
          ),
          new UIButton('$UI_Common_Cancel', btn => btn.close())
        ]
      );

      section.clearChanges();
    }
  }

  private static async generateButtonAction(section: CardSection, closeAction: () => void) {
    const fields = section.fields;
    const userCount = fields.get('UserCount');
    if (userCount < 0) {
      await showError('$KrTiles_WarnUserCountNegative');
      return false;
    }

    const partnerCount = fields.get('PartnerCount');
    if (partnerCount < 0) {
      await showError('$KrTiles_WarnPartnerCountNegative');
      return false;
    }

    if (userCount === 0 && partnerCount === 0) {
      await showMessage('$KrTiles_WarnCardsCountNotDefined');
      return false;
    }

    const text =
      userCount === 0
        ? `${localize('$KrTiles_CreatePartnersConfirmation')} ${partnerCount}. ${localize(
            '$UI_Common_ContinueConfirmation'
          )}`
        : partnerCount === 0
        ? `${localize('$KrTiles_CreateUsersConfirmation')} ${userCount}. ${localize(
            '$UI_Common_ContinueConfirmation'
          )}`
        : `${localize('$KrTiles_CreatePartnersConfirmation')} ${partnerCount}.\n` +
          `${localize('$KrTiles_CreateUsersConfirmation')} ${userCount}.\n` +
          `${localize('$UI_Common_ContinueConfirmation')}`;

    if (!(await showConfirm(text))) {
      return false;
    }

    if (
      userCount > 1000 &&
      !(await showConfirm(
        `${localize('$KrTiles_WarnTooMuchUsers')} ${userCount}. ${localize(
          '$UI_Common_ContinueConfirmation'
        )}`,
        '$UI_Common_Attention'
      ))
    ) {
      return false;
    }

    if (
      partnerCount > 1000 &&
      !(await showConfirm(
        `${localize('$KrTiles_WarnTooMuchPartners')} ${partnerCount}. ${localize(
          '$UI_Common_ContinueConfirmation'
        )}`,
        '$UI_Common_Attention'
      ))
    ) {
      return false;
    }

    closeAction();

    let response!: CardResponse;
    await showLoadingOverlay(async () => {
      const request = new CardRequest();
      request.requestType = '207e75b5-abb8-403a-a12a-897019afccf6'; // TestData
      request.info = {
        UserCount: createTypedField(userCount, DotNetType.Int32),
        PartnerCount: createTypedField(partnerCount, DotNetType.Int32)
      };
      response = await CardService.instance.request(request);
    });

    const result = response.validationResult.build();
    await showNotEmpty(result);

    return true;
  }
}

File: Source/Extensions/Tessa.Extensions.Default.Client/Tiles/KrSettingsTileExtension.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Tessa.Cards;
using Tessa.Extensions.Default.Shared;
using Tessa.Localization;
using Tessa.Platform;
using Tessa.Platform.Runtime;
using Tessa.Platform.Storage;
using Tessa.Platform.Validation;
using Tessa.UI;
using Tessa.UI.Cards;
using Tessa.UI.Tiles;
using Tessa.UI.Tiles.Extensions;

namespace Tessa.Extensions.Default.Client.Tiles
{
    /// <summary>
    /// Плитки для карточки с настройками типового решения.
    /// </summary>
    public sealed class KrSettingsTileExtension :
        TileExtension
    {
        #region Constructors

        public KrSettingsTileExtension(
            IUIHost uiHost,
            ICardRepository cardRepository,
            ISession session)
        {
            if (uiHost == null)
            {
                throw new ArgumentNullException("uiHost");
            }
            if (cardRepository == null)
            {
                throw new ArgumentNullException("cardRepository");
            }
            if (session == null)
            {
                throw new ArgumentNullException("session");
            }

            this.uiHost = uiHost;
            this.cardRepository = cardRepository;
            this.session = session;
        }

        #endregion

        #region Fields

        private readonly IUIHost uiHost;

        private readonly ICardRepository cardRepository;

        private readonly ISession session;

        #endregion

        #region Private Methods

        private void EnableOnSettingsCardAndAdministrator(object sender, TileEvaluationEventArgs e)
        {
            ICardEditorModel editor = e.CurrentTile.Context.CardEditor;

            e.SetIsEnabledWithCollapsing(
                e.CurrentTile,
                editor != null
                && editor.CardModel != null
                && editor.CardModel.CardType.ID == DefaultCardTypes.KrSettingsTypeID
                && this.session.User.IsAdministrator());
        }

        private async Task<ValidationResult> GenerateAsync(int userCount, int partnerCount)
        {
            using (TessaSplash.Create(TessaSplashMessage.CreatingMultipleCards))
            {
                CardResponse response = await this.cardRepository.RequestAsync(
                    new CardRequest
                    {
                        RequestType = DefaultRequestTypes.TestData,
                        Info =
                        {
                            { "UserCount", userCount },
                            { "PartnerCount", partnerCount },
                        }
                    });

                return response.ValidationResult.Build();
            }
        }

        private async Task GenerateButtonFuncAsync(CardSection section, Func<Task> closeActionAsync)
        {
            Dictionary<string, object> rawFields = section.RawFields;
            int userCount = rawFields.Get<int>("UserCount");
            if (userCount < 0)
            {
                TessaDialog.ShowError("$KrTiles_WarnUserCountNegative");
                return;
            }

            int partnerCount = rawFields.Get<int>("PartnerCount");
            if (partnerCount < 0)
            {
                TessaDialog.ShowError("$KrTiles_WarnPartnerCountNegative");
                return;
            }

            if (userCount == 0 && partnerCount == 0)
            {
                TessaDialog.ShowMessage("$KrTiles_WarnCardsCountNotDefined");
                return;
            }

            string text = userCount == 0
                ? (LocalizationManager.GetString("KrTiles_CreatePartnersConfirmation") + " " + partnerCount + ". " + LocalizationManager.GetString("UI_Common_ContinueConfirmation"))
                : partnerCount == 0
                    ? (LocalizationManager.GetString("KrTiles_CreateUsersConfirmation") + " " + userCount + ". " + LocalizationManager.GetString("UI_Common_ContinueConfirmation"))
                    : (LocalizationManager.GetString("KrTiles_CreatePartnersConfirmation") + " " + partnerCount + ".\r\n"
                        + LocalizationManager.GetString("KrTiles_CreateUsersConfirmation") + " " + userCount + ".\r\n"
                        + LocalizationManager.GetString("UI_Common_ContinueConfirmation"));


            //string.Format(LocalizationManager.GetString("KrTiles_CreateUsersAndPartnersConfirmation"), userCount, partnerCount);

            if (!TessaDialog.Confirm(text))
            {
                return;
            }

            if (userCount > 1000
                && !TessaDialog.Confirm(
                    string.Format(LocalizationManager.GetString("KrTiles_WarnTooMuchUsers") + " " + userCount + LocalizationManager.GetString("UI_Common_ContinueConfirmation")),
                    "$UI_Common_Attention"))
            {
                return;
            }

            if (partnerCount > 1000
                && !TessaDialog.Confirm(
                    //KrTiles_WarnTooMuchPartners
                    string.Format(LocalizationManager.GetString("KrTiles_WarnTooMuchPartners") + " " + partnerCount + LocalizationManager.GetString("UI_Common_ContinueConfirmation")),
                    //string.Format("Количество создаваемых контрагентов {0} слишком велико. Продолжить?", partnerCount),
                    "$UI_Common_Attention"))
            {
                return;
            }

            await closeActionAsync();

            ValidationResult result = await this.GenerateAsync(userCount, partnerCount);
            TessaDialog.ShowNotEmpty(result);
        }

        #endregion

        #region Command Actions

        private async void GenerateActionAsync(object parameter)
        {
            ICardEditorModel editor = UIContext.Current.CardEditor;
            ICardModel model;
            if (editor != null
                && (model = editor.CardModel) != null)
            {
                CardSection section = model.Card.Sections["KrCardGeneratorVirtual"];
                CardTypeNamedForm form = editor.CardModel.CardType.Forms
                    .First(x => string.Equals(x.Name, "Generator", StringComparison.Ordinal));

                await this.uiHost.ShowFormDialogAsync(
                    form.TabCaption,
                    form,
                    editor.CardModel,
                    async (f, ct) =>
                    {
                        IDictionary<string, object> fields = section.Fields;
                        fields["UserCount"] = 0;
                        fields["PartnerCount"] = 0;
                    },
                    buttons: new[]
                    {
                        new UIButton(
                            LocalizationManager.GetString("KrTiles_CreateCardsButton"),
                            async btn => await this.GenerateButtonFuncAsync(section, btn.CloseAsync),
                            isDefault: true),

                        new UIButton(LocalizationManager.GetString("UI_Common_Cancel"), isCancel: true),
                    });

                section.ClearChanges();
            }
        }

        #endregion

        #region Base Overrides

        public override Task InitializingGlobal(ITileGlobalExtensionContext context)
        {
            ITilePanel panel = context.Workspace.LeftPanel;
            panel.Tiles.Add(
                new Tile(
                    TileNames.GenerateTestCards,
                    TileHelper.SplitCaption("$KrTiles_GenerateTestCards"),
                    context.Icons.Get("Thin1"),
                    panel,
                    new DelegateCommand(this.GenerateActionAsync),
                    TileGroups.Top,
                    order: 1,
                    verticalAlignment: TileVerticalAlignment.Bottom,
                    evaluating: this.EnableOnSettingsCardAndAdministrator));

            return Task.CompletedTask;
        }

        #endregion
    }
}
